src/efftester.ml

```diff
@@ -98,15 +102,29 @@ type lit =
   | LitBool of bool
   | LitStr of string
   | LitList of lit list
+  | LitOption of lit option
```

Comment:
You don't need `LitOption` here, because we have the `list_intro_rules`. For example, `Some 1` can already be represented as `Constructor (Option Int, "Some", Lit (LitInt 1))`, no need for `LitOption (Some (LitInt 1))` in addition.

Reply:
I created the `LitOption` variant because if `literal_gen` is required to generate a
literal of type `(int option) list`, it fails because it does not have access to type
`term` and outer permute generator. My proposition below solves this problem without
introducing `LitOption`.

Currently, lists in our generator can only contain literals (`type lit`). Representation
of options (`Constructor _`) in efftester is contained in type `term`. 

I propose to move lists representation from literals to terms. 

```ocaml
type term = 
      ...
    | List of term list * eff
```

Thus our list would be able to contain not only literals but also more intricate
expressions. It would also need to track effects present. 

This move would entail introducing `list_intro_rules` (I am not sure about
`list_elim_rules` yet). These changes go out of scope of this PR, but without it the tester remains buggy. 

Let me know what you think about it.

Reply:
I think that's an excellent idea, and I think that it is fine to not have `list_elim_rules` yet (more on this below). To generate a list with elements `e1, e2, e3`, we have two options (unless the list is empty and there is only `[]`): `e1::e2::e3::[]` and `[e1; e2; e3]`. Here is what I would say:

- if we had decided to integrate lists as ADTs, we would represent them using `Constructor` instead of a dedicated expression constructor, and there the natural representation would be `e1 :: ...`. 
- With the `List of term list * eff` that you propose, I think the natural choice is `[e1; e2; ...]`. Note that this choice extends very naturally to generation of Array, with an `Array of term list * eff` that behaves in the same way and generates `[| e1; e2; ... |]` (let's keep arrays out for now).

Regarding list_elim: we could decide to generate pattern-matching in this case, something like `match foo with [] -> ... | x::xs -> ...`. That would be natural if we decided to see lists as any other ADT, and build them using `Constructor`, but this is not what you proposed above -- I think it is fine to not do it. With the current approach, the way we use lists in generated programs is by using `List.hd` and `List.tl`, which are in the initial environment. Once we have support for options, we could go further by introducing a `List.pop_opt : 'a list -> ('a * 'a list) option` in the environment, which would allow us to generate programs that check for the empty list and return a different value in that case (using pattern-matching on the resulting option value).

Reply:
Would you be weighted toward a repr that is a `term list * eff` or `Constructor`?

Reply:
`term list * eff` is the simplest to implement, I think, so I would personally go for that. I don't expect many bugs in OCaml implementations to come from the handling of lists, which tend to be well-tested.

Note: what exactly is the problem if `literal_gen` cannot generate a term of type `(int option) list`? Naively I would expect the term generator to cleanly handle a failure of literal_gen; after all, not all types have literals, the same problem occurred with `(int -> int) list` before you added the Option type. Should `literal_gen` returns an `option` type to indicate that no term generation is possible at certain types?

Reply:
Below in the code you can see that if the type `t` is `int option list`, it is first matched with `List t` and if `size <> 0`, then an arbitrary-sized list is generated again using `literal_gen` in order to generate elements of the list. As the list can only generate literal-element list, `literal_gen` failed to generate `int option` elements for the list. 

```ocaml
let rec literal_gen t eff size =
    match t with
    | Unit -> Gen.return LitUnit
    | Int -> Gen.map (fun i -> LitInt i) int_gen
    | Float -> Gen.map (fun f -> LitFloat f) float_gen
    | Bool -> Gen.map (fun b -> LitBool b) Gen.bool
    | String -> Gen.map (fun s -> LitStr s) string_gen
    | Option _t' -> (* FIXME: implement *) failwith "literal_gen: not implemented"
    | List (Typevar _) -> Gen.return (LitList [])
    | List t ->
      if size = 0
      then Gen.return (LitList [])
      else
        Gen.map
          (fun ls -> LitList ls)
          (Gen.list_size (Gen.int_bound (sqrt size)) (literal_gen t eff (sqrt size)))
    (*     (Gen.list_size (Gen.int_bound (size/2)) (literal_gen t eff (size/2))) *)
    (* FIXME: - one element should/can have effect, if 'eff' allows *)
    (*        - list items should be able to contain arbitrary effectful exps *)
    | Typevar _ -> failwith "literal_gen: typevar arg. should not happen"
    | Fun _ -> failwith "literal_gen: funtype arg. should not happen"
  ;;
```

We indeed could use an option return from the `literal_gen`, but we anyway wanted to move list generation as a separate rule, right?

Reply:
Yes, I think moving list generation as a separate rule is the right direction. We should never call `literal_gen` if the input type is not a base type, which is the simpler way to get rid of the problem.